### PR TITLE
Remove the "exit(rc)" of the onnx-mlir compile interface

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
Converted the "exit" in the onnx-mlir compile interface to report error codes. I also smoothed the OnnxMlirCompiler interface to return the compiled file name, which is useful as there is no "fix" rules to determine the names on all platforms.

How names were formed was also in lots of places in the code. Now its all centralized in one new function
`std::string getTargetFilename`. This function is now thoroughly used within the code and tests.

I also added an example in  docs/doc_example that uses the C/C++ interface to compile and run an example. 

The plan was to add the doc, but to do the doc, I started doing the example, and then I saw issues in the interface. So docs will be added.

### CODE REVIEW

File: docs/doc_example/main.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+int main() {
+  // Read compiler options from environment first, and ensure -O3 is used.
+  int rc;
+  std::cout << "Start compiling add.onnx" << std::endl;
+  rc = onnx_mlir::omSetCompilerOptionsFromEnv("TEST_DOC_EXAMPLE");
+  if (rc != onnx_mlir::CompilerSuccess) {
+    std::cerr << "Failed to process TEST_DOC_EXAMPLE compiler options with "
+                 "error code "
+              << rc << "." << std::endl;
+    return 1;
```
Comment (etiotto) [inline docs/doc_example/main.cpp:16]: Suggest to return the appropriate enumerator defined in OMCompilerTypes.h

File: docs/doc_example/main.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  std::cout << "Compiled succeeded with results in file: " << libFilename
+            << std::endl;
+
+  // Prepare the execution session and get input signature.
+  onnx_mlir::ExecutionSession *session;
+  try {
+    session = new onnx_mlir::ExecutionSession(libFilename);
+  } catch (const std::runtime_error &error) {
+    std::cerr << "error while creating execution session: " << error.what()
+              << std::endl;
```
Comment (etiotto) [inline docs/doc_example/main.cpp:47]: See comment al line 16.

File: src/Compiler/CompilerUtils.cpp [hunk 3]
```diff
@@ -239,19 +242,21 @@ struct Command {
                    << "Program path: " << _path << "\n"
                    << "Command execution failed."
                    << "\n";
-      exit(rc);
+      return rc;
     }
 
     // Restore saved work directory.
     llvm::sys::fs::set_current_path(cur_wdir);
+    return 0;
```
Comment (etiotto) [inline src/Compiler/CompilerUtils.cpp:250]: return CompilerSuccess;

File: src/Runtime/ExecutionSession.cpp [hunk 4]
```diff
@@ -32,6 +34,15 @@ const std::string ExecutionSession::_outputSignatureName = "omOutputSignature";
...
 ExecutionSession::ExecutionSession(
     std::string sharedLibPath, bool defaultEntryPoint) {
+
+  // For some reason, the call to get the permanent library requires a path, and
+  // when none is provided, it still does not look in the current directory. To
+  // remediate this situation, if no path is given, we add a "./" to the path to
+  // force looking in the current directory.
+  if (llvm::sys::path::parent_path(sharedLibPath).empty()) {
+    sharedLibPath = "./" + sharedLibPath;
+    std::cout << "hi alex, added local " << sharedLibPath << std::endl;
```
Comment (etiotto) [inline src/Runtime/ExecutionSession.cpp:44]: remove stray trace.

File: src/Compiler/OnnxMlirCompiler.cpp [hunk 5]
```diff
@@ -51,38 +51,59 @@ ONNX_MLIR_EXPORT const char *omGetCompilerOption(const OptionKind kind) {
...
     if (errorMessage != NULL)
       *errorMessage = strdup(internalErrorMessage.c_str());
     return rc;
   }
-  return compileModule(module, context, outputBaseName, emissionTarget);
+  rc = compileModule(module, context, outputBaseName, emissionTarget);
+  if (rc == CompilerSuccess && outputFilename) {
+    // Copy Filename
+    *outputFilename =
+        strdup(getTargetFilename(outputBaseName, emissionTarget).c_str());
```
[the remark this answers is not in the document]
Comment (gongsu832) [reply src/Compiler/OnnxMlirCompiler.cpp:75]: Not sure what you mean by `strip(outputFilenameStr)`.

File: src/Compiler/CompilerUtils.cpp [hunk 6]
```diff
@@ -221,7 +223,8 @@ struct Command {
     llvm::sys::fs::make_absolute(cur_wdir, new_wdir);
     if (std::error_code ec = llvm::sys::fs::set_current_path(new_wdir)) {
       llvm::errs() << StringRef(new_wdir).str() << ": " << ec.message() << "\n";
-      exit(ec.value());
+      assert(ec.value() != 0 && "Expected nonnull error return value");
```
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Compiler/CompilerUtils.cpp:226]: Since we have an error, and its relying on `ec` being non null, but then return `ec.value()`, which must be nonzero to be then seen as an error, it seems ok to assert that it is nonzero. I will put a explanation.

File: docs/doc_example/main.cpp (hunk 2 above)
Comment (AlexandreEichenberger) [reply docs/doc_example/main.cpp:47]: ok

File: src/Compiler/CompilerUtils.cpp (hunk 3 above)
Comment (AlexandreEichenberger) [reply src/Compiler/CompilerUtils.cpp:250]: I can, but error code are both from execution and compilation. Keeping 0 is fine I believe.

File: src/Runtime/ExecutionSession.cpp (hunk 4 above)
Comment (AlexandreEichenberger) [reply src/Runtime/ExecutionSession.cpp:44]: Oh no, lost my chance at posterity fame :-)

File: src/Compiler/OnnxMlirCompiler.cpp [hunk 7]
```diff
@@ -51,38 +51,59 @@ ONNX_MLIR_EXPORT const char *omGetCompilerOption(const OptionKind kind) {
 
 ONNX_MLIR_EXPORT int64_t omCompileFromFile(const char *inputFilename,
     const char *outputBaseName, EmissionTargetType emissionTarget,
-    const char **errorMessage) {
+    char **outputFilename, char **errorMessage) {
```
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Compiler/OnnxMlirCompiler.cpp:54]: yes, tx

File: docs/doc_example/main.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  }
+  rc = onnx_mlir::omSetCompilerOption(onnx_mlir::CompilerOptLevel, "3");
+  if (rc != onnx_mlir::CompilerSuccess) {
+    std::cerr << "Failed to process -O3 compiler option with error code " << rc
+              << "." << std::endl;
+    return 1;
+  }
+  // Compile the doc example into a model library.
+  char *errorMessage;
+  char *compiledFilename;
```
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply docs/doc_example/main.cpp:26]: ok

File: docs/doc_example/main.cpp [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  int rc;
+  std::cout << "Start compiling add.onnx" << std::endl;
+  rc = onnx_mlir::omSetCompilerOptionsFromEnv("TEST_DOC_EXAMPLE");
+  if (rc != onnx_mlir::CompilerSuccess) {
+    std::cerr << "Failed to process TEST_DOC_EXAMPLE compiler options with "
+                 "error code "
+              << rc << "." << std::endl;
+    return 1;
+  }
+  rc = onnx_mlir::omSetCompilerOption(onnx_mlir::CompilerOptLevel, "3");
```
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply docs/doc_example/main.cpp:18]: both seem to work, used your suggestion.

File: src/Compiler/OnnxMlirCompiler.cpp [hunk 10]
```diff
@@ -51,38 +51,59 @@ ONNX_MLIR_EXPORT const char *omGetCompilerOption(const OptionKind kind) {
...
+  if (errorMessage)
+    *errorMessage = NULL;
+  if (outputFilename)
+    *outputFilename = NULL;
```
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Compiler/OnnxMlirCompiler.cpp:58]: changed it to remove the setting to null, though I don't feel comfortable with that solution, at least for the error message. Let me know if you continue to think its not needed

File: src/Compiler/OnnxMlirCompiler.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (gongsu832) [reply src/Compiler/OnnxMlirCompiler.cpp]: You mean `*outputFilename = strdup(outputFilenameStr.c_str());`? I'm not a C++ expert but it sounds to me this doesn't make much difference. My understanding is that string assignment in C++ is a copy assignment, i.e., when you do
```
string outputFilenameStr = getTargetFilename(outputBaseName, emissionTarget);
```
`outputFilenameStr` gets a copy of what `getTargetFilename` returns (calls this returned string `foo`). So the question still remains: when `foo` goes out of scope?

I was wondering if you could just do
```
*outputFilename = getTargetFilename(outputBaseName, emissionTarget).c_str();
```
without doing a strdup (then have to free it by the caller). But if `foo` goes out of scope when `omCompileFromFile` returns, then strdup is necessary and the original code is fine.

File: src/Runtime/ExecutionSession.cpp [hunk 11]
```diff
@@ -32,6 +34,15 @@ const std::string ExecutionSession::_outputSignatureName = "omOutputSignature";
...
+  }
   _sharedLibraryHandle =
```
[the remark this answers is not in the document]
Comment (gongsu832) [reply src/Runtime/ExecutionSession.cpp:45]: I don't think the two are directly comparable. `omCompileFromFile` is a compile time function whereas `ExecutionSession` is a runtime function. So it's not a problem that they might have different search behavior. But the main point is that we shouldn't be doing things under the cover to change the search behavior not expected by the caller.

File: src/Compiler/OnnxMlirCompiler.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Compiler/OnnxMlirCompiler.cpp]: >  But if foo goes out of scope when omCompileFromFile returns, then strdup is necessary and the original code is fine.

Exactly, it is a parameter that is returned to the function caller scope.

File: src/Compiler/CompilerUtils.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (gongsu832) [reply src/Compiler/CompilerUtils.cpp:229]: Looking at how `llvm::sys::fs::set_current_path` is implemented,
```
 std::error_code set_current_path(const Twine &path) {
   SmallString<128> path_storage;
   StringRef p = path.toNullTerminatedStringRef(path_storage);
  
   if (::chdir(p.begin()) == -1)
     return std::error_code(errno, std::generic_category());
  
   return std::error_code();
 }
```
you see that for both error or no error cases, it returns a std::error_code(). So the correct check should be,
```
std::error_code ec = llvm::sys::fs::set_current_path(new_wdir);
if (ec.value()) {
  llvm::errs() << StringRef(new_wdir).str() << ": " << ec.message() << "\n";
  return ec.value();
}
```

File: src/Compiler/OnnxMlirCompiler.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Compiler/OnnxMlirCompiler.cpp]: code does not set any of the const **char to NULL anymore

File: src/Compiler/CompilerUtils.cpp (hunk 3 above)
Comment (etiotto) [reply src/Compiler/CompilerUtils.cpp:235]: ok

File: src/Runtime/ExecutionSession.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (etiotto) [reply src/Runtime/ExecutionSession.cpp:45]: For a compile section I would expect to search for the file to compile in the current directory. And if not found emit an error. I don't think the compile interface need to change.

[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Runtime/ExecutionSession.cpp:45]: > For a compile section I would expect to search for the file to compile in the current directory. And if not found emit an error. I don't think the compile interface need to change.

@etiotto so would't we expect the same for loading the .so file when starting an execution session? 

Or maybe the right way to do it is that If the compiler interface returns the .so with a fully qualified path, so that it flows naturally from the compiler to the execution session, and I don't need to add the "special casing" in the loading of the file to artificially look in the "./" too, since that does not sit well with @gongsu832 . What do you both think about this?

[the remark this answers is not in the document]
Comment (gongsu832) [reply src/Runtime/ExecutionSession.cpp:45]: @AlexandreEichenberger 

> ```c++
>   rc = onnx_mlir::omCompileFromFile("add.onnx", "add-cppinterface",
>       onnx_mlir::EmitLib, &compiledFilename, &errorMessage);
> ```
> succeed without the ./ because LLVM seems to search the current dir, so this lead to a strange situation where the path for the compilation could skip ./ but the execution session did not.

What file(s) are you referring to when you say "succeed without the ./ because LLVM seems to search the current dir"?

As mentioned before, compilation search by LLVM has it own rules, and runtime search is not done by LLVM but rather ld.so, the dynamic loader, which has its own rules. So I'm not sure why it is confusing when the behaviors are different.

[the remark this answers is not in the document]
Comment (etiotto) [reply src/Runtime/ExecutionSession.cpp:45]: Agree with @gongsu832. The compilation command:
```
gcc -shared -o libfoo.so foo.c 
```
searches for the input file `foo.c` in the current directory (that is without having to specify ./foo.c).

Conversely to run the executable one needs to provide the relative path (that is prepend the exec with a dot: ./bar). 
Also the executable depends on a shared library which will only be found if the directory containing the shared library is in the LD_LIBRARY_PATH). 

In summary it make sense to me that the compilation and execution environment do not need to behave in the same way.

**Review by etiotto — APPROVED** [review]
LGTM

File: src/Runtime/ExecutionSession.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Runtime/ExecutionSession.cpp:45]: The issue is that we don't have the ability to add a `-L` option programmatically at this point.
